### PR TITLE
Add support to LOBSTER >=5.0 

### PR DESCRIPTION
**Pull request by naik-aakash**
## Summary

This PR aims to update the `lobster.io.inputs`  and `lobster.io.outputs` modules to support LOBSTER v5. There are some changes in v5.1, mainly with the format of output files. Also, new output files are generated from fragment orbital analysis. A few new keywords have also been added.  

# Todos

- [x] add new keywords from lobster >= v5 to the inputs module
- [x] update ICOHPLIST parser
- [x] update GROSSPOP parser
- [x] update Lobsterout parser
- [x] Add parser for POLARIZATION.lobster lobster file (LOBSTER >=5.0 generates this)
- [x] Add parser for BWDF.lobster lobster file
- [x] Add support for `COXXCAR.LCFO.lobster, ICOXXLIST.LCFO.lobster, GROSSPOP.LCFO.lobster` file parsing
- [x] Add support for `DOSCAR.LCFO.lobster, CHARGE.LCFO.lobster` file parsing
- [x] Add attributes to lobsterout parser for MOFE generated files
- [x] Add tests

### CODE REVIEW

**Comment by JaGeo** [comment]
@naik-aakash thanks! I was about to start doing it myself 🙈

**Comment by JaGeo** [comment]
I think we also don't have a parser for the polarization file 😅. But maybe, this can be done in a subsequent pull request

**Comment by JaGeo** [comment]
@naik-aakash should we maybe get a part of this merged as a bug fix and do the other things later?

**Comment by QuantumChemist** [comment]
> @naik-aakash should we maybe get a part of this merged as a bug fix and do the other things later?

I was in contact with the LOBSTER developers and they will release a hotfix this week for a bug that prevents the LCFO anaylsis currently. I would suggest to wait with merging until then.

**Comment by naik-aakash** [comment]
> @naik-aakash should we maybe get a part of this merged as a bug fix and do the other things later?

We can do this. I can add test for this updates in next 1 or 2 days. 

Also, a bug was found in the new MOFE feature. Thus, I could not generate test files or write parsers for outputs for that specific analysis. In anycase a new version for LOBSTER is expected soon.

**Comment by naik-aakash** [comment]
Hi @JaGeo, this PR can now be reviewed. I am happy to address any issues that are spotted or needs some change.

File: src/pymatgen/io/lobster/outputs.py [hunk 1]
```diff
@@ -855,9 +912,16 @@ def get_structure_with_charges(self, structure_filename: PathLike) -> Structure:
             Structure Object with Mulliken and Loewdin charges as site properties.
         """
         struct = Structure.from_file(structure_filename)
-        mulliken = self.mulliken
-        loewdin = self.loewdin
-        site_properties = {"Mulliken Charges": mulliken, "Loewdin Charges": loewdin}
+        site_properties = {}
+        if "LCFO" not in Path(self._filename).name:
```
Comment (JaGeo) [inline src/pymatgen/io/lobster/outputs.py:916]: Not sure this is really a safe check. Also, not sure if fragment orbital charges can ever be assigned to a specific site. I would throw an error.

File: src/pymatgen/io/lobster/outputs.py [hunk 2]
```diff
@@ -1640,12 +1738,52 @@ def __init__(
...
+                        Spin.up: float(cleanlines[3]),
+                        Spin.down: float(cleanlines[4]),
+                    }
+                    small_dict["Loewdin GP"][cleanlines[2]] = {
+                        Spin.up: float(cleanlines[5]),
+                        Spin.down: float(cleanlines[6]),
+                    }
 
-                elif len(cleanlines) > 0:
+                elif len(cleanlines) > 0 and "spin" not in line and "LCFO" in Path(self._filename).name:
```
Comment (JaGeo) [inline src/pymatgen/io/lobster/outputs.py:1776]: Is there another way to check this? Not sure that the filename alone is safe enough

File: src/pymatgen/io/lobster/outputs.py [hunk 3]
```diff
@@ -1661,16 +1799,23 @@ def get_structure_with_total_grosspop(self, structure_filename: PathLike) -> Str
...
-        loewdin_gps: list[dict] = []
-        for grosspop in self.list_dict_grosspop:
-            mulliken_gps.append(grosspop["Mulliken GP"]["total"])
-            loewdin_gps.append(grosspop["Loewdin GP"]["total"])
-
-        site_properties = {
-            "Total Mulliken GP": mulliken_gps,
-            "Total Loewdin GP": loewdin_gps,
-        }
+        site_properties = {}
```
Comment (JaGeo) [inline src/pymatgen/io/lobster/outputs.py:1802]: Again: not sure adding thr LCFO prop to a site is even wanted

**Review by JaGeo — COMMENTED** [review]
I left some thoughts. Overall, nice work 😃

File: src/pymatgen/io/lobster/outputs.py (hunk 1 above)
Comment (naik-aakash) [reply src/pymatgen/io/lobster/outputs.py:916]: Have updated the method to raise a ValueError now for LCFO files.

File: src/pymatgen/io/lobster/outputs.py (hunk 2 above)
Comment (naik-aakash) [reply src/pymatgen/io/lobster/outputs.py:1776]: Thanks, updated the code to check if it is atom or molecule type from the file data itself. 

However, I still rely on filename checks in Grosspop.get_structure_with_total_grosspop / Charge.get_structure_with_charges methods. If I want to do this check using file data, I would need to define another argument to the function and set an attribute, `_file_type` to the class to keep it MSONable.

File: src/pymatgen/io/lobster/outputs.py (hunk 3 above)
Comment (naik-aakash) [reply src/pymatgen/io/lobster/outputs.py:1802]: Raises value error now

**Comment by naik-aakash** [comment]
Hi @JaGeo, I have now addressed all the review comments where possible. For the `Cohpcar `and `Doscar`  parsers, I do not see any other way to extract file types from the data within the file itself, so I have left them as they are, and it seems to work for now.

**Comment by JaGeo** [comment]
So far, we mostly let users choose the filetype (eg Icohps, Icoops). I would do the same here.

**Comment by naik-aakash** [comment]
> So far, we mostly let users choose the filetype (eg Icohps, Icoops). I would do the same here.

Oh, okay. This way sounds much simpler. I have now added the `is_lcfo` arg to the relevant parsers with the docstrings.

**Review by JaGeo — APPROVED** [review]
This looks great @naik-aakash ! Thank you!

@shyuep @mkhorton this would be ready to merge from my side. Please let us know if we should change anything else. Thank you. The test that fails currently is related to matplotlib and windows and I don't think this has anything to do with our pull request.

**Comment by naik-aakash** [comment]
Hi @shyuep , now all tests also pass after last merge. If there are no specific comments on the changes, it is ready to be merged.

**Comment by shyuep** [comment]
Great. Thanks!

**Pull request merged**